File: cpp/src/ErrorHandler.cpp
#include "treon/ErrorHandler.hpp"
#include "treon/Constants.hpp"

#include <QDebug>
#include <QDesktopServices>
#include <QUrl>
#include <QDateTime>

using namespace treon;

// MARK: - TreonException Implementation

TreonException::TreonException(ErrorType type, const QString& message)
    : m_type(type)
    , m_message(message.isEmpty() ? TreonException::getUserFriendlyMessage(type) : message)
{
}

TreonException::TreonException(const QString& message)
    : m_type(ErrorType::UnknownError)
    , m_message(message)
{
}

QStringList TreonException::recoveryActions() const {
    return TreonException::getRecoveryActions(m_type);
}

bool TreonException::isRecoverable() const {
    return TreonException::isErrorRecoverable(m_type);
}

void TreonException::raise() const {
    throw *this;
}

QException* TreonException::clone() const {
    return new TreonException(*this);
}

QString TreonException::getUserFriendlyMessage(ErrorType type) {
    switch (type) {
        case ErrorType::FileNotFound: return ErrorMessages::fileNotFound;
        case ErrorType::InvalidJSON: return ErrorMessages::invalidJSON;
        case ErrorType::FileTooLarge: return ErrorMessages::fileTooLarge;
        case ErrorType::UnsupportedFileType: return ErrorMessages::unsupportedFileType;
        case ErrorType::PermissionDenied: return ErrorMessages::permissionDenied;
        case ErrorType::CorruptedFile: return ErrorMessages::corruptedFile;
        case ErrorType::NetworkError: return ErrorMessages::networkError;
        case ErrorType::UserCancelled: return ErrorMessages::userCancelled;
        case ErrorType::InvalidURL: return ErrorMessages::invalidURL;
        case ErrorType::CurlCommandFailed: return ErrorMessages::curlCommandFailed;
        case ErrorType::EmptyPasteboard: return ErrorMessages::emptyPasteboard;
        case ErrorType::LoadingFailed: return ErrorMessages::loadingFailed;
        case ErrorType::UnknownError: return ErrorMessages::unknownError;
    }
    return ErrorMessages::unknownError;
}

QStringList TreonException::getRecoveryActions(ErrorType type) {
    QStringList actions;
    actions << "Cancel";
    
    if (isErrorRecoverable(type)) {
        actions.prepend("Retry");
    }
    
    switch (type) {
        case ErrorType::PermissionDenied:
            actions << "Settings";
            break;
        case ErrorType::NetworkError:
        case ErrorType::UnknownError:
            actions << "Contact Support";
            break;
        default:
            break;
    }
    
    return actions;
}

bool TreonException::isErrorRecoverable(ErrorType type) {
    switch (type) {
        case ErrorType::FileNotFound:
        case ErrorType::PermissionDenied:
        case ErrorType::NetworkError:
        case ErrorType::UnknownError:
            return true;
        case ErrorType::UserCancelled:
        case ErrorType::InvalidJSON:
        case ErrorType::FileTooLarge:
        case ErrorType::UnsupportedFileType:
        case ErrorType::CorruptedFile:
        case ErrorType::InvalidURL:
        case ErrorType::CurlCommandFailed:
        case ErrorType::EmptyPasteboard:
        case ErrorType::LoadingFailed:
            return false;
    }
    return false;
}

// MARK: - ErrorHandler Implementation

ErrorHandler::ErrorHandler(QObject* parent)
    : QObject(parent)
{
}

void ErrorHandler::handleError(const TreonException& exception) {
    qDebug() << "Error occurred:" << exception.message();
    
    const QString message = exception.message();
    const QStringList actions = exception.recoveryActions();
    
    updateErrorState(message, actions);
    emit errorOccurred(message, exception.type());
}

void ErrorHandler::handleError(ErrorType type, const QString& message) {
    TreonException exception(type, message);
    handleError(exception);
}

void ErrorHandler::handleError(const QString& message) {
    TreonException exception(message);
    handleError(exception);
}

void ErrorHandler::clearError() {
    if (!m_currentError.isEmpty()) {
        m_currentError.clear();
        m_recoveryActions.clear();
        emit currentErrorChanged();
        emit hasErrorChanged();
        emit recoveryActionsChanged();
    }
}

void ErrorHandler::performRecoveryAction(ErrorRecoveryAction action) {
    switch (action) {
        case ErrorRecoveryAction::Retry:
            // Retry logic would be implemented based on context
            clearError();
            break;
        case ErrorRecoveryAction::Cancel:
        case ErrorRecoveryAction::Ignore:
            clearError();
            break;
        case ErrorRecoveryAction::OpenSettings:
            // Open system preferences
            QDesktopServices::openUrl(QUrl("x-apple.systempreferences:com.apple.preference.security?Privacy"));
            clearError();
            break;
        case ErrorRecoveryAction::ContactSupport:
            // Open email client
            {
                const QString subject = "Treon Support Request";
                const QString body = QString("Error Details:\n- App Version: %1\n- Build: %2\n- Error: %3\n- Timestamp: %4\n\nPlease describe what you were doing when this error occurred:\n\n")
                    .arg(AppConstants::version)
                    .arg(AppConstants::buildNumber)
                    .arg(m_currentError)
                    .arg(QDateTime::currentDateTime().toString());
                
                const QString mailtoUrl = QString("mailto:%1?subject=%2&body=%3")
                    .arg(AppConstants::supportEmail)
                    .arg(QUrl::toPercentEncoding(subject))
                    .arg(QUrl::toPercentEncoding(body));
                
                QDesktopServices::openUrl(QUrl(mailtoUrl));
            }
            clearError();
            break;
    }
}

QString ErrorHandler::getUserFriendlyMessage(ErrorType type) const {
    switch (type) {
        case ErrorType::FileNotFound: return ErrorMessages::fileNotFound;
        case ErrorType::InvalidJSON: return ErrorMessages::invalidJSON;
        case ErrorType::FileTooLarge: return ErrorMessages::fileTooLarge;
        case ErrorType::UnsupportedFileType: return ErrorMessages::unsupportedFileType;
        case ErrorType::PermissionDenied: return ErrorMessages::permissionDenied;
        case ErrorType::CorruptedFile: return ErrorMessages::corruptedFile;
        case ErrorType::NetworkError: return ErrorMessages::networkError;
        case ErrorType::UserCancelled: return ErrorMessages::userCancelled;
        case ErrorType::InvalidURL: return ErrorMessages::invalidURL;
        case ErrorType::CurlCommandFailed: return ErrorMessages::curlCommandFailed;
        case ErrorType::EmptyPasteboard: return ErrorMessages::emptyPasteboard;
        case ErrorType::LoadingFailed: return ErrorMessages::loadingFailed;
        case ErrorType::UnknownError: return ErrorMessages::unknownError;
    }
    return ErrorMessages::unknownError;
}

QStringList ErrorHandler::getRecoveryActions(ErrorType type) const {
    QStringList actions;
    actions << "Cancel";
    
    if (isErrorRecoverable(type)) {
        actions.prepend("Retry");
    }
    
    switch (type) {
        case ErrorType::PermissionDenied:
            actions << "Settings";
            break;
        case ErrorType::NetworkError:
        case ErrorType::UnknownError:
            actions << "Contact Support";
            break;
        default:
            break;
    }
    
    return actions;
}

bool ErrorHandler::isErrorRecoverable(ErrorType type) const {
    switch (type) {
        case ErrorType::FileNotFound:
        case ErrorType::PermissionDenied:
        case ErrorType::NetworkError:
        case ErrorType::UnknownError:
            return true;
        default:
            return false;
    }
}

void ErrorHandler::updateErrorState(const QString& message, const QStringList& actions) {
    m_currentError = message;
    m_recoveryActions = actions;
    emit currentErrorChanged();
    emit hasErrorChanged();
    emit recoveryActionsChanged();
}

File: Treon/Core/ErrorHandler.swift
//
//  ErrorHandler.swift
//  Treon
//
//  Created by Vivek on 2025-10-19.
//  Copyright © 2025 Treon. All rights reserved.
//

import Foundation
import SwiftUI
import OSLog
import Combine

// MARK: - Error Handling Protocol
protocol ErrorHandling {
    func handleError(_ error: Error, context: String?)
    func showErrorAlert(_ error: Error, context: String?)
    func logError(_ error: Error, context: String?)
}

// MARK: - Error Recovery Actions
enum ErrorRecoveryAction {
    case retry
    case cancel
    case ignore
    case openSettings
    case contactSupport
}

// MARK: - Error Context
struct ErrorContext {
    let operation: String
    let userAction: String?
    let timestamp: Date
    let additionalInfo: [String: Any]?

    init(operation: String, userAction: String? = nil, additionalInfo: [String: Any]? = nil) {
        self.operation = operation
        self.userAction = userAction
        self.timestamp = Date()
        self.additionalInfo = additionalInfo
    }
}

// MARK: - Enhanced Error Handler
class TreonErrorHandler: ObservableObject, ErrorHandling {
    @Published var currentError: Error?
    @Published var errorMessage: String = ""
    @Published var showErrorAlert: Bool = false
    @Published var isRecoverable: Bool = false
    @Published var recoveryActions: [ErrorRecoveryAction] = []

    private let logger = Loggers.error

    // MARK: - Error Handling Methods

    func handleError(_ error: Error, context: String? = nil) {
        logError(error, context: context)
        if isRunningUnderTests() {
            updateStateSynchronously(error: error, context: context)
            postErrorNotification(error: error, context: context)
            return
        }
        DispatchQueue.main.async { [weak self] in
            guard let self else { return }
            self.updateStateAsynchronously(error: error, context: context)
            self.postErrorNotification(error: error, context: context)
        }
    }

    private func isRunningUnderTests() -> Bool {
        ProcessInfo.processInfo.environment["XCTestConfigurationFilePath"] != nil
    }

    private func updateStateSynchronously(error: Error, context: String?) {
        currentError = error
        errorMessage = getUserFriendlyMessage(for: error)
        isRecoverable = isErrorRecoverable(error)
        recoveryActions = getRecoveryActions(for: error)
        showErrorAlert = true
    }

    private func updateStateAsynchronously(error: Error, context: String?) {
        currentError = error
        errorMessage = getUserFriendlyMessage(for: error)
        isRecoverable = isErrorRecoverable(error)
        recoveryActions = getRecoveryActions(for: error)
        showErrorAlert = true
    }

    private func postErrorNotification(error: Error, context: String?) {
        NotificationCenter.default.post(
            name: NotificationNames.errorOccurred,
            object: error,
            userInfo: ["context": context ?? "Unknown"]
        )
    }

    func showErrorAlert(_ error: Error, context: String? = nil) {
        handleError(error, context: context)
    }

    func logError(_ error: Error, context: String? = nil) {
        let contextString = context ?? "Unknown"
        logger.error("Error in \(contextString): \(error.localizedDescription)")

        // Log additional details for debugging
        if let fileManagerError = error as? FileManagerError {
            logger.error("FileManagerError details: \(fileManagerError)")
        }

        // Log to console for development
        #if DEBUG
        logger.error("🚨 Error in \(contextString): \(error)")
        if let nsError = error as NSError? {
            logger.error("   Domain: \(nsError.domain)")
            logger.error("   Code: \(nsError.code)")
            logger.error("   UserInfo: \(nsError.userInfo)")
        }
        #endif
    }

    // MARK: - Error Recovery

    func performRecoveryAction(_ action: ErrorRecoveryAction) {
        switch action {
        case .retry:
            retryLastOperation()
        case .cancel:
            dismissError()
        case .ignore:
            dismissError()
        case .openSettings:
            openAppSettings()
        case .contactSupport:
            contactSupport()
        }
    }

    func dismissError() {
        if ProcessInfo.processInfo.environment["XCTestConfigurationFilePath"] != nil {
            self.currentError = nil
            self.errorMessage = ""
            self.showErrorAlert = false
            self.isRecoverable = false
            self.recoveryActions = []
        } else {
            DispatchQueue.main.async {
                self.currentError = nil
                self.errorMessage = ""
                self.showErrorAlert = false
                self.isRecoverable = false
                self.recoveryActions = []
            }
        }
    }

    // MARK: - Private Methods

    private func getUserFriendlyMessage(for error: Error) -> String {
        if let fileManagerError = error as? FileManagerError {
            return fileManagerError.errorDescription ?? ErrorMessages.unknownError
        }

        // Handle common system errors
        if let nsError = error as NSError? {
            switch nsError.domain {
            case NSCocoaErrorDomain:
                return handleCocoaError(nsError)
            case NSURLErrorDomain:
                return handleURLError(nsError)
            case NSPOSIXErrorDomain:
                return handlePOSIXError(nsError)
            default:
                return error.localizedDescription.isEmpty ? ErrorMessages.unknownError : error.localizedDescription
            }
        }

        return error.localizedDescription.isEmpty ? ErrorMessages.unknownError : error.localizedDescription
    }

    private func handleCocoaError(_ error: NSError) -> String {
        switch error.code {
        case NSFileReadNoSuchFileError:
            return ErrorMessages.fileNotFound
        case NSFileReadNoPermissionError:
            return ErrorMessages.permissionDenied
        case NSFileReadCorruptFileError:
            return ErrorMessages.corruptedFile
        case NSFileReadTooLargeError:
            return ErrorMessages.fileTooLarge
        default:
            return error.localizedDescription
        }
    }

    private func handleURLError(_ error: NSError) -> String {
        switch error.code {
        case NSURLErrorNotConnectedToInternet:
            return "No internet connection available"
        case NSURLErrorTimedOut:
            return "Request timed out"
        case NSURLErrorCannotFindHost:
            return "Cannot find server"
        case NSURLErrorCannotConnectToHost:
            return "Cannot connect to server"
        case NSURLErrorNetworkConnectionLost:
            return "Network connection lost"
        default:
            return ErrorMessages.networkError
        }
    }

    private func handlePOSIXError(_ error: NSError) -> String {
        switch error.code {
        case Int(ENOENT):
            return ErrorMessages.fileNotFound
        case Int(EACCES):
            return ErrorMessages.permissionDenied
        case Int(EIO):
            return ErrorMessages.corruptedFile
        default:
            return error.localizedDescription
        }
    }

    private func isErrorRecoverable(_ error: Error) -> Bool {
        if let fileManagerError = error as? FileManagerError {
            switch fileManagerError {
            case .fileNotFound, .permissionDenied, .networkError:
                return true
            case .userCancelled, .invalidJSON, .fileTooLarge, .unsupportedFileType, .corruptedFile:
                return false
            case .unknownError:
                return true
            }
        }

        if let nsError = error as NSError? {
            switch nsError.domain {
            case NSURLErrorDomain:
                return true // Network errors are usually recoverable
            case NSCocoaErrorDomain:
                switch nsError.code {
                case NSFileReadNoSuchFileError, NSFileReadNoPermissionError:
                    return true
                default:
                    return false
                }
            case NSPOSIXErrorDomain:
                switch nsError.code {
                case Int(ENOENT), Int(EACCES):
                    return true
                default:
                    return false
                }
            default:
                return false
            }
        }

        return false
    }

    private func getRecoveryActions(for error: Error) -> [ErrorRecoveryAction] {
        var actions: [ErrorRecoveryAction] = [.cancel]

        if isErrorRecoverable(error) {
            actions.insert(.retry, at: 0)
        }

        if let fileManagerError = error as? FileManagerError {
            switch fileManagerError {
            case .permissionDenied:
                actions.append(.openSettings)
            case .networkError, .unknownError:
                actions.append(.contactSupport)
            default:
                break
            }
        }

        return actions
    }

    private func retryLastOperation() {
        // This would be implemented based on the specific operation that failed
        // For now, we'll just dismiss the error
        dismissError()
    }

    private func openAppSettings() {
        // Open system preferences for the app
        if let url = URL(string: "x-apple.systempreferences:com.apple.preference.security?Privacy") {
            NSWorkspace.shared.open(url)
        }
    }

    private func contactSupport() {
        // Open email client with support information
        let subject = "Treon Support Request"
        let body = """
        Error Details:
        - App Version: \(AppConstants.version)
        - Build: \(AppConstants.buildNumber)
        - Error: \(currentError?.localizedDescription ?? "Unknown")
        - Timestamp: \(Date())

        Please describe what you were doing when this error occurred:

        """

        let mailtoURL = "mailto:\(AppConstants.supportEmail)?subject=\(subject.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? "")&body=\(body.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? "")"

        if let url = URL(string: mailtoURL) {
            NSWorkspace.shared.open(url)
        }
    }
}

// MARK: - Error Alert View
struct ErrorAlertView: View {
    @ObservedObject var errorHandler: TreonErrorHandler
    let onDismiss: () -> Void

    var body: some View {
        VStack(spacing: UIConstants.mediumSpacing) {
            // Error Icon
            Image(systemName: "exclamationmark.triangle.fill")
                .font(.system(size: 48))
                .foregroundColor(UIConstants.errorRed)

            // Error Title
            Text("Error")
                .font(.headline)
                .foregroundColor(.primary)

            // Error Message
            Text(errorHandler.errorMessage)
                .font(.body)
                .foregroundColor(.secondary)
                .multilineTextAlignment(.center)
                .fixedSize(horizontal: false, vertical: true)

            // Recovery Actions
            if errorHandler.isRecoverable {
                VStack(spacing: UIConstants.smallSpacing) {
                    Text("What would you like to do?")
                        .font(.caption)
                        .foregroundColor(.secondary)

                    HStack(spacing: UIConstants.smallSpacing) {
                        ForEach(errorHandler.recoveryActions, id: \.self) { action in
                            Button(actionTitle(for: action)) {
                                errorHandler.performRecoveryAction(action)
                                onDismiss()
                            }
                            .buttonStyle(.bordered)
                        }
                    }
                }
            } else {
                Button("OK") {
                    errorHandler.dismissError()
                    onDismiss()
                }
                .buttonStyle(.borderedProminent)
            }
        }
        .padding(UIConstants.largeSpacing)
        .frame(maxWidth: 400)
    }

    private func actionTitle(for action: ErrorRecoveryAction) -> String {
        switch action {
        case .retry:
            return "Retry"
        case .cancel:
            return "Cancel"
        case .ignore:
            return "Ignore"
        case .openSettings:
            return "Settings"
        case .contactSupport:
            return "Contact Support"
        }
    }
}

// MARK: - Error Handling View Modifier
struct ErrorHandlingModifier: ViewModifier {
    @StateObject private var errorHandler = TreonErrorHandler()

    func body(content: Content) -> some View {
        content
            .environmentObject(errorHandler)
            .alert("Error", isPresented: $errorHandler.showErrorAlert) {
                ForEach(errorHandler.recoveryActions, id: \.self) { action in
                    Button(actionTitle(for: action)) {
                        errorHandler.performRecoveryAction(action)
                    }
                }
                Button("Cancel") {
                    errorHandler.dismissError()
                }
            } message: {
                Text(errorHandler.errorMessage)
            }
    }

    private func actionTitle(for action: ErrorRecoveryAction) -> String {
        switch action {
        case .retry:
            return "Retry"
        case .cancel:
            return "Cancel"
        case .ignore:
            return "Ignore"
        case .openSettings:
            return "Settings"
        case .contactSupport:
            return "Contact Support"
        }
    }
}

// MARK: - View Extension
extension View {
    func withErrorHandling() -> some View {
        self.modifier(ErrorHandlingModifier())
    }
}
